File: src/day01/main.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

int main() {
    FILE *f = fopen("./puzzle_input/day01", "r");
    if (f==NULL) {
        printf("Error opening input");
        return 1;
    }
    uint32_t p1first = 0;
    uint32_t p1last = 0;
    uint32_t p1total = 0;
    uint32_t p2first = 0;
    uint32_t p2last = 0;
    uint32_t p2total = 0;
    char buf[23000] = {0};
    char *read_head = buf;
    int num_read = fread(&buf, 1, sizeof(buf), f);
    char *end = buf + num_read;
    while (read_head < end) {
        if (*read_head == '\n') {
            p1total += p1first * 10 + p1last;
            p2total += p2first * 10 + p2last;
            p1first = p1last = p2first = p2last = 0;
            ++read_head;
        } else if (('1' <= *read_head) && (*read_head <= '9')) {
            p1last = p2last = *read_head - '0';
            ++read_head;
        } else if (0 == strncmp(read_head, "one", 3)) {
            p2last = 1;
            read_head += 3;
        } else if (0 == strncmp(read_head, "two", 3)) {
            p2last = 2;
            read_head += 3;
        } else if (0 == strncmp(read_head, "three", 5)) {
            p2last = 3;
            read_head += 5;
        } else if (0 == strncmp(read_head, "four", 4)) {
            p2last = 4;
            read_head += 4;
        } else if (0 == strncmp(read_head, "five", 4)) {
            p2last = 5;
            read_head += 4;
        } else if (0 == strncmp(read_head, "six", 3)) {
            p2last = 6;
            read_head += 3;
        } else if (0 == strncmp(read_head, "seven", 5)) {
            p2last = 7;
            read_head += 5;
        } else if (0 == strncmp(read_head, "eight", 5)) {
            p2last = 8;
            read_head += 5;
        } else if (0 == strncmp(read_head, "nine", 4)) {
            p2last = 9;
            read_head += 4;
        } else {
            ++read_head;
        }
        if (p1first == 0) {
            p1first = p1last;
        }
        if (p2first == 0) {
            p2first = p2last;
        }
    }
    p1total += p1first * 10 + p1last;
    p2total += p2first * 10 + p2last;
    printf("part 1: %d\n", p1total);
    printf("part 2: %d\n", p2total);
    return 0;
}

File: src/day01/hs/main.hs
-- ghc .\src\day01\hs\main.hs ; .\src\day01\hs\main
import Data.Char (digitToInt, isDigit)
import Data.List (isPrefixOf)

p1Line :: String -> Int
p1Line line = a * 10 + b
  where
    digits = [ digitToInt c | c <- line, isDigit c ]
    a = head digits
    b = last digits

p2Digits:: String -> [Int]
p2Digits [] = []
p2Digits ('0':s) = 0 : p2Digits s
p2Digits ('1':s) = 1 : p2Digits s
p2Digits ('2':s) = 2 : p2Digits s
p2Digits ('3':s) = 3 : p2Digits s
p2Digits ('4':s) = 4 : p2Digits s
p2Digits ('5':s) = 5 : p2Digits s
p2Digits ('6':s) = 6 : p2Digits s
p2Digits ('7':s) = 7 : p2Digits s
p2Digits ('8':s) = 8 : p2Digits s
p2Digits ('9':s) = 9 : p2Digits s
p2Digits s
  | "one" `isPrefixOf` s = 1 : p2Digits (drop 3 s)
  | "two" `isPrefixOf` s = 2 : p2Digits (drop 3 s)
  | "three" `isPrefixOf` s = 3 : p2Digits (drop 5 s)
  | "four" `isPrefixOf` s = 4 : p2Digits (drop 4 s)
  | "five" `isPrefixOf` s = 5 : p2Digits (drop 4 s)
  | "six" `isPrefixOf` s = 6 : p2Digits (drop 3 s)
  | "seven" `isPrefixOf` s = 7 : p2Digits (drop 5 s)
  | "eight" `isPrefixOf` s = 8 : p2Digits (drop 5 s)
  | "nine" `isPrefixOf` s = 9 : p2Digits (drop 4 s)
  | otherwise = p2Digits (drop 1 s)

p2Line :: String -> Int
p2Line line = a * 10 + b
  where
    digits = p2Digits line
    a = head digits
    b = last digits

p1 :: [String] -> Int
p1 lines = sum $ map p1Line lines

p2 :: [String] -> Int
p2 lines = sum $ map p2Line lines

main :: IO ()
main = do
  input <- readFile "puzzle_input/day01"
  putStrLn $ ("part 1: " ++) $ show $ p1 $ lines input
  putStrLn $ ("part 2: " ++) $ show $ p2 $ lines input